os/kernel/binary_manager/binary_manager_recovery.c

```diff
@@ -62,6 +63,38 @@ static void binary_manager_board_reset(void)
 #endif
 }
 
+static void recovery_release_binary_sem(int binid)
+{
+	sem_t *sem;
+	irqstate_t flags;
+	FAR struct semholder_s *holder;
+
+	flags = irqsave();
+
+	sem = (sem_t *)sq_peek(&g_sem_list);
+	while (sem) {
+#if CONFIG_SEM_PREALLOCHOLDERS > 0
+		for (holder = sem->hhead; holder; holder = holder->flink)
+#else
+		holder = &sem->holder;
+#endif
+		{
+			if (holder && holder->htcb && holder->htcb->group && holder->htcb->group->tg_loadtask == binid) {
+				/* It doesn't release if the semaphore with zero value and semcount is larger than or equal with 0. */
+				if (IS_ZERO_SEM(sem) && sem->semcount >= 0) {
+					sem = sq_next(sem);
```

Comment:
In this case, sem will be changed to next, but holder is not.
Please check this logic.

Reply:
right. I modified.